File: apps/satellite-service/src/routes/health.ts
/**
 * @file routes/health.ts
 * @description Health check endpoint.
 * Returns 200 when the service is ready to handle traffic, with
 * sub-system status for readiness probes in Kubernetes.
 */

import { Router, type Request, type Response } from 'express';

import { prisma } from '../server';
import { getRedis } from '../lib/redis';
import logger from '../lib/logger';

export const healthRouter = Router();

interface HealthStatus {
  status: 'ok' | 'degraded' | 'unhealthy';
  timestamp: string;
  service: string;
  version: string;
  uptime: number;
  checks: {
    postgres: 'ok' | 'error';
    redis: 'ok' | 'error';
  };
}

/**
 * GET /health
 * Liveness probe — always returns 200 if the process is running.
 */
healthRouter.get('/', (_req: Request, res: Response) => {
  res.status(200).json({ status: 'ok', timestamp: new Date().toISOString() });
});

/**
 * GET /health/ready
 * Readiness probe — checks connectivity to PostgreSQL and Redis.
 * Returns 200 only when all dependencies are healthy.
 */
healthRouter.get('/ready', async (_req: Request, res: Response) => {
  const checks: HealthStatus['checks'] = {
    postgres: 'error',
    redis: 'error',
  };

  // Check PostgreSQL
  try {
    await prisma.$queryRaw`SELECT 1`;
    checks.postgres = 'ok';
  } catch (err) {
    logger.warn(err, 'Health check: PostgreSQL failed');
  }

  // Check Redis
  try {
    const pong = await getRedis().ping();
    if (pong === 'PONG') checks.redis = 'ok';
  } catch (err) {
    logger.warn(err, 'Health check: Redis failed');
  }

  const allOk = Object.values(checks).every((v) => v === 'ok');
  const status: HealthStatus['status'] = allOk ? 'ok' : 'degraded';

  const health: HealthStatus = {
    status,
    timestamp: new Date().toISOString(),
    service: 'satellite-service',
    version: process.env['npm_package_version'] ?? '0.0.0',
    uptime: process.uptime(),
    checks,
  };

  res.status(allOk ? 200 : 503).json(health);
});

File: apps/planetary-service/src/routers/health.py
"""Health check router for Kubernetes liveness and readiness probes."""

from __future__ import annotations

import time
from typing import Literal

from fastapi import APIRouter
from pydantic import BaseModel

from ..cache import get_redis
from ..database import get_engine
from ..config import settings

router = APIRouter(tags=["Health"])

START_TIME = time.time()


class HealthResponse(BaseModel):
    status: Literal["ok", "degraded", "unhealthy"]
    service: str
    version: str
    uptime_seconds: float
    timestamp: str
    checks: dict[str, Literal["ok", "error"]] = {}


@router.get("/health", summary="Liveness probe")
async def liveness() -> dict[str, str]:
    """Always returns 200 if the process is alive."""
    from datetime import UTC, datetime
    return {"status": "ok", "timestamp": datetime.now(UTC).isoformat()}


@router.get("/health/ready", response_model=HealthResponse, summary="Readiness probe")
async def readiness() -> HealthResponse:
    """Returns 200 only when PostgreSQL and Redis are reachable."""
    from datetime import UTC, datetime
    import asyncio

    checks: dict[str, Literal["ok", "error"]] = {}

    # Check Redis
    try:
        redis = await get_redis()
        await redis.ping()
        checks["redis"] = "ok"
    except Exception:
        checks["redis"] = "error"

    # Check PostgreSQL
    try:
        engine = get_engine()
        async with engine.connect() as conn:
            await conn.execute(__import__("sqlalchemy").text("SELECT 1"))
        checks["postgres"] = "ok"
    except Exception:
        checks["postgres"] = "error"

    all_ok = all(v == "ok" for v in checks.values())

    return HealthResponse(
        status="ok" if all_ok else "degraded",
        service="planetary-service",
        version=settings.VERSION,
        uptime_seconds=round(time.time() - START_TIME, 2),
        timestamp=datetime.now(UTC).isoformat(),
        checks=checks,
    )
